_posts/2020-01-23-scatter-gather-pt2.md

```diff
@@ -0,0 +1,142 @@
+
+In most cases, using a persistence option with pessimistic locking capabilities will prevent major problems for you, at the cost of slower throughput at some times. However, there is another approach which doesn't impact the endpoint throughput and also won't cause concurrency exceptions and retries. It's a lot more complex than just using pessimistic locking but sometimes the throughput limitations can be an important factor or you might rely on a persistence not support pessimistic locking. Here's a rough overview how this approach will work:
+
+1. Instead of aggregating the results in the saga directly, a simple message handler will receive the response message type.
+2. The handler will store the response data in an append-only storage which allows extremely fast inserts without concurrency violations.
+3. A separate process will repeatedly check if all expected responses arrived and will continue the business process managed by the saga.
+4. The saga will aggregate all the response data.
+
+Let's look at some sample code:
+
+### Scattering
+
+The `ScatterGatherSaga` will dispatch a large amount of messages (e.g. 500) when the saga is started:
+
+```
+class ScatterGatherSaga : Saga<ScatterGatherSagaData>, 
+    IAmStartedByMessages<StartSagaCommand>
+{
+    public async Task Handle(StartSagaCommand message, IMessageHandlerContext context)
+    {
+        Data.NumberOfMessages = message.NumberOfMessages;
+
+        for (int i = 0; i < message.NumberOfMessages; i++)
+        {
+            await context.Send(new RequestResponseCommand()
+            {
+                BatchId = message.Id,
+                RequestId = i
+            });
+        }
+    }
+
+    // mapping omitted
+}
+
+```
+
+Another endpoint will respond to each `RequestResponseCommand` with a `ResponseMessage`. Instead of handling those directly as part of the saga, we implement a stateless message handler:
+
+### The handler
+
+The handler will receive the `ResponseMessage` and store the results in an append-only manner into a database (e.g. MonoDB in this case):
+
+
+```
+class GatherHandler : IHandleMessages<ResponseMessage>
+{
+    public async Task Handle(ResponseMessage message, IMessageHandlerContext context)
+    {
+        var mongoSession = context.SynchronizedStorageSession.GetClientSession();
+        var mongoCollection = mongoSession.Client.GetDatabase(Program.DatabaseName)
+            .GetCollection<ResponseResult>("results");
+
+        await mongoCollection
+            .InsertOneAsync(mongoSession, new ResponseResult
+            {
+                Id = context.MessageId,
+                Result = message.RequestResult,
+                SagaId = message.SagaId
```

Comment:
You're not sending out the saga id above

Reply:
yeah it's not really the saga id, need to rename the property. Good catch!